### DISCUSSION TITLE
Mehr als 10 Wechselrichter...

### DISCUSSION THREAD

**Discussion by bynight66** [tbnobody/OpenDTU#1217] [General]
Was mache ich wenn ich mehr als 10 Wechselrichter habe?

**Comment by MN-237** [comment]
Habe das selbe Problem!
Ich nutze 12x HM-1500 Hoeymiles, leider werden nur 9 Stück in den "Wechselrichter Einstellungen" im Web Interface angezeigt.
Habe die aktuelle Version [v23.8.8] gerade neu kompiliert und in der Configuration.h die "INV_MAX_COUNT" von 10 auf 20 geändert.
Leider ohne Wirkung.

**Reply by DPO99** [reply]
Nunja...es werden 10 Stück pro DTU erlaubt. Lösung -> 2. DTU ;)

**Reply by bynight66** [reply]
Warum nur 10 und nicht 99 oder 100??? Meine DTU Pro zeigt mindestens 99 an und auch EINE Gesamtleistung. Das wird ja bei mehreren DTUs auch schwierig... Ne Programmierer Laune?
Ahoy unterstützt übrigens 16 Wechselrichter. Reicht aber auch nicht. :-(

**Reply by bynight66** [reply]
steht irgendwo daß die Änderung von "INV_MAX_COUNT" von 10 auf 20 helfen soll oder war das nur so eine Idee?

**Reply by DPO99** [reply]
Das wird schon seine Gründe haben. Der Entwickler macht das sicher nicht um uns zu ärgern.....

**Reply by bynight66** [reply]
Glaub ich nicht... Wenn 10 gehen gehen auch 11... 
16 wären eine natürliche Grenze oder 32... Ahoy bietet 16 und die Hoymiles DTU 99... Also technisch sollte es möglich sein...

**Reply by DPO99** [reply]
Naja tut mir leid aber wenn Du alles besser weißt dann mach es doch selbst? Es gab schon mehrfach die Diskussion und auch dass es teilweise Richtung 10 WR schon zu Problemen kommen kann. Und wenn Dir diese Lösung nicht zusagt dann nutz eben das Original für den 20 fachen Preis. :)

**Reply by bynight66** [reply]
Der Preis war nie das Kriterium. Das Original nutze ich schon seit nem Jahr. Funktioniert gut, auch mit >20 HMs... kommuniziert aber wohl nicht mit Victron. Nun suche ich Alternativen die ebenfalls mit >20 HMs kommunizieren können. "Mach es doch selbst" ist da nicht hilfreich.

**Reply by oldendeti1** [reply]
Hallo, ich betreibe 10 HM (8*1500 und 2*600). Die werden auch alle mit den Erträgen und so angezeigt, allerdings kann ich im Menü "Wechselrichter" den 10. WR nicht mehr ansprechen bzw. einstellen. Ich kann mich entsinnen, dass ich auch mit früheren Versionen dieses Problem hatte, ich aber zwischenzeitlich die WR über dieses Menü ansprechen konnte. Wäre schön, wenn zumindest die Gründe für die Einschränkung kommuniziert würden, dann bräuchten Zänkereien (s. o.) nicht stattzufinden. Und solange ich die Ergebnisse aller WR auch über Mqtt erhalte, ist es mir auch egal. Aber wissen würde ich es doch schon gerne...

**Reply by MN-237** [reply]
Genau so wie du es beschreibst ist es bei mir auch! Ich habe momentan von den 12 WR 10 Aktiv (für die beiden anderen brauche ich noch weitere PV Panels), diese werden auch sauber in "Live-Daten" dargestellt. Nur in "Einstellungen --> Wechselrichter" werden davon NUR 9 angezeigt.
Wie gesagt das Kompilieren mit "INV_MAX_COUNT" auf 20 gesetzt hat dieses Problem nicht gelöst. Habe mir den gesamten Quellcode angesehen, habe aber keinen weiteren Code gefunden außer die Variable "INV_MAX_COUNT" der die Anzahl der WR beeinflusst.

**Reply by tbnobody** [reply]
Es ist immer ein Trade-Off zwischen Features und Anzahl Inverter. 

10 Inverter sind so gewählt das es mehr sind als bei einer DTU-Lite und beim RAM auch noch etwas Luft nach oben verfügbar ist (für zukünftige Features). Wenn jemand meint den Source Code so optimieren zu können damit 20 oder gar 100 oder noch besser 255 Inverter klappen (bei ganzen 320kb RAM des ESP32) bin ich sehr auf das Ergebnis gespannt. OpenDTU-OnBattery (Fork) hat die Inverter Anzahl schon auf 5 reduziert da durch die Ansteuerung der Batterieregler usw. nicht genug RAM zur Verfügung steht.

Wenn der 10. Inverter in der Liste nicht bearbeitbar ist, würde ich bitten kurz den Output von `http://[OPENDTU_IP]/api/inverter/list` zu posten. Ggf. ist hier der Puffer für den JSON Rendering Output etwas zu klein gewählt.

**Reply by MN-237** [reply]
Danke für die Info mit dem JSON Puffer, könnte gut möglich sein das es daran liegt!
Muss ich nochmal wenn ich Zeit habe den Quellcode durchforsten nach dem JSON Rendering Output Puffer.

Wie gewünscht anbei der Output von "http://[OPENDTU_IP]/api/inverter/list" (ich habe meine Inverter Serials durch "X" ersetzt):

{"inverter":[{"id":0,"name":"Garage RECHTS","order":0,"serial":"XXXXXXXXXXXX","poll_enable":true,"poll_enable_night":true,"command_enable":true,"command_enable_night":true,"type":"HM-1000, HM-1200, HM-1500","channel":[{"name":"GANZ RECHTS","max_power":385,"yield_total_offset":0},{"name":"unused","max_power":0,"yield_total_offset":0},{"name":"Garage (r) Dach (links unten)","max_power":420,"yield_total_offset":0},{"name":"Garage (r) Dach (links oben)","max_power":420,"yield_total_offset":0}]},{"id":1,"name":"Garage MITTE","order":0,"serial":"XXXXXXXXXXXX","poll_enable":true,"poll_enable_night":true,"command_enable":true,"command_enable_night":true,"type":"HM-1000, HM-1200, HM-1500","channel":[{"name":"Garage Dach hinten ganz rechts","max_power":420,"yield_total_offset":0},{"name":"Garage Dach hinten rechts","max_power":420,"yield_total_offset":0},{"name":"LINKS MITTE","max_power":385,"yield_total_offset":0},{"name":"RECHTS MITTE","max_power":385,"yield_total_offset":0}]},{"id":2,"name":"Garage LINKS","order":0,"serial":"XXXXXXXXXXXX","poll_enable":true,"poll_enable_night":true,"command_enable":true,"command_enable_night":true,"type":"HM-1000, HM-1200, HM-1500","channel":[{"name":"Garage Dach hinten ganz links","max_power":420,"yield_total_offset":0},{"name":"Garage Dach hinten links","max_power":420,"yield_total_offset":0},{"name":"RECHTS LINKS","max_power":385,"yield_total_offset":0},{"name":"GANZ LINKS","max_power":385,"yield_total_offset":0}]},{"id":3,"name":"Garage DACH","order":0,"serial":"XXXXXXXXXXXX","poll_enable":true,"poll_enable_night":true,"command_enable":true,"command_enable_night":true,"type":"HM-1000, HM-1200, HM-1500","channel":[{"name":"Garage Dach vorne ganz links","max_power":420,"yield_total_offset":0},{"name":"Garage Dach vorne links","max_power":420,"yield_total_offset":0},{"name":"Garage Dach vorne rechts","max_power":420,"yield_total_offset":0},{"name":"Garage Dach vorne ganz rechts","max_power":420,"yield_total_offset":0}]},{"id":4,"name":"Haus West","order":0,"serial":"XXXXXXXXXXXX","poll_enable":true,"poll_enable_night":true,"command_enable":true,"command_enable_night":true,"type":"HM-1000, HM-1200, HM-1500","channel":[{"name":"","max_power":0,"yield_total_offset":0},{"name":"","max_power":0,"yield_total_offset":0},{"name":"LINKS","max_power":385,"yield_total_offset":0},{"name":"RECHTS","max_power":385,"yield_total_offset":0}]},{"id":5,"name":"Dach VORNE","order":0,"serial":"XXXXXXXXXXXX","poll_enable":true,"poll_enable_night":true,"command_enable":true,"command_enable_night":true,"type":"HM-1000, HM-1200, HM-1500","channel":[{"name":"rechts aussen","max_power":385,"yield_total_offset":0},{"name":"rechts innen","max_power":385,"yield_total_offset":0},{"name":"links innen","max_power":385,"yield_total_offset":0},{"name":"links aussen","max_power":385,"yield_total_offset":0}]},{"id":6,"name":"Dach HINTEN","order":0,"serial":"XXXXXXXXXXXX","poll_enable":true,"poll_enable_night":true,"command_enable":true,"command_enable_night":true,"type":"HM-1000, HM-1200, HM-1500","channel":[{"name":"rechts aussen","max_power":385,"yield_total_offset":0},{"name":"rechts innen","max_power":385,"yield_total_offset":0},{"name":"links aussen","max_power":385,"yield_total_offset":0},{"name":"unused","max_power":0,"yield_total_offset":0}]},{"id":7,"name":"Garage DACH R","order":0,"serial":"XXXXXXXXXXXX","poll_enable":true,"poll_enable_night":true,"command_enable":true,"command_enable_night":true,"type":"HM-1000, HM-1200, HM-1500","channel":[{"name":"unused","max_power":0,"yield_total_offset":0},{"name":"Garage Dach RECHTS rechts oben","max_power":420,"yield_total_offset":0},{"name":"Garage Dach RECHTS rechts unte","max_power":420,"yield_total_offset":0},{"name":"unused","max_power":0,"yield_total_offset":0}]},{"id":8,"name":"Anbau MITTE","order":0,"serial":"XXXXXXXXXXXX","poll_enable":true,"poll_enable_night":true,"command_enable":true,"command_enable_night":true,"type":"HM-1000, HM-1200, HM-1500","channel":[{"name":"Anbau (rechts oben)","max_power":420,"yield_total_offset":0},{"name":"Anbau (mitte oben)","max_power":420,"yield_total_offset":0},{"name":"Anbau (links oben)","max_power":420}]}]}

**Reply by tbnobody** [reply]
Wie vermutet ist der Puffer das Problem. Wird heute noch eine neue Version geben in der das behoben sein sollte.

**Reply by MN-237** [reply]
Super, werde ich testen und berichten!

**Reply by MN-237** [reply]
Habe testweise mal "JSON_BUFFER_SIZE" auf 16383 erhöht, Problem besteht weiterhin!

**Reply by tbnobody** [reply]
JSON_BUFFER_SIZE gilt ja auch nur für die Config und nicht für die Json response...

**Reply by oldendeti1** [reply]
Da möchte ich mich aber mal ganz herzlich beim Entwickler bedanken! Und auch mein Kompliment ausdrücken zu dieser tollen Software. Schöne Grüße aus Dortmund vom oldendti1 😀

**Reply by MN-237** [reply]
Vielen Dank an die Entwickler von OpenDTU!
Die Anzeige der "Wechselrichter Einstellungen" funktioniert jetzt!

**Reply by bynight66** [reply]
@tbnobody Ok, das Problem ist also der Speicher des ESP32... Danke für die Erklärung! :-) Lösungen mit anderen Mikrocontrollern mit mehr Speicher gibt es wohl noch nicht? Oder hat das einfach noch niemand gebraucht?

**Comment by desiue** [comment]
Hallo zusammen,
ich habe das Problem das ich mir einen 11en Wechselrichter gekauft habe, eigentlich aus dem Grund weil 6 Solarmodule den HMT-2250-6T Wechselrichter ans Clipping gebracht haben. 
Jetzt habe ich leider Feststellen müssen "Es werden nur 10 Wechselrichter unterstützt!"

Hm also müsste ich jetzt eine weitere OpenDTU aufbauen nur für einen weiteren Wechselrichter.
Hinzu kommen weitere Anpassungen in der Heimautomation und das Problem das dann die Langzeit Daten aus influxdb in Grafana auch nicht mehr zusammen passen weil sich der Name ändert. Alles sehr unschön...

Alles in Allem ein Riesenaufwand. GIbt es eine Möglichkeit eine Firmware zu bekommen die 11 WR kann?

Kann das jemand bitte bitte kompilieren?

Von der Hardware her handelt sich um einen ESP32-D0WD-V3 mit einem CMT2300A FW Version [v26.3.30] 
Aktuell sind 10 WR dran und ein Abfrageintervall von 2 Sekunden auf 868,25Mhz die Systemressourcen scheinen nicht das Problem zu sein.



<img width="1101" height="1820" alt="dtu" src="https://github.com/user-attachments/assets/a21ee303-9478-4dfa-891a-547b61ca3ee3" />


Ok meine Verzweiflung war groß genug, vscode + Platfomio plugin + git installiert.
In OpenDTU/include/Configuration.h  #define INV_MAX_COUNT 11 geändert, compiliert und geflasht. 
Konnte auch den 11ten WR hinzufügen. 

Läuft! danke für die Software!!!
